File: continue-on-failure-skips-compensation/python/reconcile_skipped_compensation.py
"""Reconcile Medusa orders left partial by a continueOnPermanentFailure step.

.config({ continueOnPermanentFailure: true }) opts a step out of the saga's rollback
contract. Per Medusa's docs, the compensation function of the flagged step will not
be called, and the workflow keeps running subsequent steps as if nothing happened.
If that step already committed a side effect, an order, a captured payment, or a
reservation, and a later step then fails and triggers a rollback, the orchestrator
still does not retroactively undo the flagged step's work (PR #12027, issue #11266).

This lists recent orders with payments and fulfillments expanded, classifies each
one with a pure function, and reports every orphan as a structured record for a
human to triage. The only guarded write is deleting a dangling reservation that has
no live order line. Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/medusa/continue-on-failure-skips-compensation/
"""
import os
import json
import logging
import datetime
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reconcile_skipped_compensation")

BACKEND_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
ADMIN_EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
ADMIN_PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
SINCE_HOURS = float(os.environ.get("SINCE_HOURS", "24"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

CAPTURED_STATUSES = {"captured", "authorized"}
STUCK_FULFILLMENT_STATUSES = {"not_fulfilled"}
DELETABLE = {"orphaned_reservation_no_order_line"}


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def admin_get(token, path, params=None):
    r = requests.get(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        params=params or {},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def admin_delete(token, path):
    r = requests.delete(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def classify_orphan(order, failed_steps):
    """Pure decision function. No I/O.

    order: {"id": str, "payment_status": str, "fulfillment_status": str,
            "payments": [{"status": str}], "fulfillments": [any], "items": [any]}
    failed_steps: list of {"action": str, "handlerType": "invoke" | "compensate"}

    Returns "orphaned_payment_no_fulfillment" | "orphaned_reservation_no_order_line" | "ok".
    """
    has_continue_on_failure = any(
        s.get("handlerType") == "invoke" and "continueOnPermanentFailure" in s.get("action", "")
        for s in failed_steps
    )

    payment_committed = order.get("payment_status") in CAPTURED_STATUSES and bool(order.get("payments"))
    fulfillment_missing = (
        order.get("fulfillment_status") in STUCK_FULFILLMENT_STATUSES
        and not order.get("fulfillments")
    )
    if payment_committed and fulfillment_missing and has_continue_on_failure:
        return "orphaned_payment_no_fulfillment"

    has_dangling_reservation = any(
        s.get("action") == "reserveInventoryStep" and s.get("handlerType") == "invoke"
        for s in failed_steps
    )
    if has_dangling_reservation and not order.get("items"):
        return "orphaned_reservation_no_order_line"

    return "ok"


def list_recent_orders(token, since_iso):
    orders = []
    offset = 0
    limit = 100
    while True:
        data = admin_get(token, "/admin/orders", {
            "fields": "id,status,fulfillment_status,payment_status,*payments,*fulfillments,*items",
            "created_at[$gte]": since_iso,
            "limit": limit,
            "offset": offset,
        })
        orders.extend(data["orders"])
        offset += limit
        if offset >= data["count"]:
            return orders


def failed_steps_for_order(order):
    """Placeholder hook: in a real deployment, load the failed-step trail for this
    order's workflow transaction (for example from your own audit log of
    { result, errors } from someWorkflow(container).run({ input, throwOnError: false })).
    Returns [] when there is nothing on file, which classify_orphan treats as "ok".
    """
    return order.get("_failed_steps", [])


def run():
    token = get_admin_token()
    since_iso = (
        datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(hours=SINCE_HOURS)
    ).isoformat()
    orders = list_recent_orders(token, since_iso)

    reported = 0
    cleaned = 0
    for order in orders:
        failed_steps = failed_steps_for_order(order)
        outcome = classify_orphan(order, failed_steps)
        if outcome == "ok":
            continue

        if outcome in DELETABLE:
            reservation_id = order.get("_dangling_reservation_id")
            log.warning(
                "Order %s classified as %s. reservation_id=%s. %s",
                order["id"], outcome, reservation_id, "Would delete" if DRY_RUN else "Deleting",
            )
            if not DRY_RUN and reservation_id:
                admin_delete(token, f"/admin/reservations/{reservation_id}")
            cleaned += 1
        else:
            record = {
                "order_id": order["id"],
                "action": next((s["action"] for s in failed_steps if s.get("handlerType") == "invoke"), None),
                "error_message": next((s.get("message") for s in failed_steps), None),
                "classification": outcome,
                "reported_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
            }
            log.info("Orphan detected: %s", json.dumps(record))
            reported += 1

    log.info(
        "Done. %d order(s) reported for human triage, %d reservation(s) %s.",
        reported, cleaned, "to delete" if DRY_RUN else "deleted",
    )


if __name__ == "__main__":
    run()

File: continue-on-failure-skips-compensation/node/reconcile-skipped-compensation.js
/**
 * Reconcile Medusa orders left partial by a continueOnPermanentFailure step.
 *
 * .config({ continueOnPermanentFailure: true }) opts a step out of the saga's rollback
 * contract. Per Medusa's docs, the compensation function of the flagged step will not
 * be called, and the workflow keeps running subsequent steps as if nothing happened.
 * If that step already committed a side effect, an order, a captured payment, or a
 * reservation, and a later step then fails and triggers a rollback, the orchestrator
 * still does not retroactively undo the flagged step's work (PR #12027, issue #11266).
 *
 * This lists recent orders with payments and fulfillments expanded, classifies each
 * one with a pure function, and reports every orphan as a structured record for a
 * human to triage. The only guarded write is deleting a dangling reservation that has
 * no live order line. Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/continue-on-failure-skips-compensation/
 */
import { pathToFileURL } from "node:url";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const SINCE_HOURS = Number(process.env.SINCE_HOURS || 24);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const CAPTURED_STATUSES = new Set(["captured", "authorized"]);
const STUCK_FULFILLMENT_STATUSES = new Set(["not_fulfilled"]);
const DELETABLE = new Set(["orphaned_reservation_no_order_line"]);

/**
 * Pure decision function. No I/O.
 *
 * @param {{ id: string, payment_status: string, fulfillment_status: string, payments: {status:string}[], fulfillments: unknown[], items?: unknown[] }} order
 * @param {{ action: string, handlerType: "invoke" | "compensate" }[]} failedSteps
 * @returns {"orphaned_payment_no_fulfillment" | "orphaned_reservation_no_order_line" | "ok"}
 */
export function classifyOrphan(order, failedSteps) {
  const hasContinueOnFailure = failedSteps.some(
    (s) => s.handlerType === "invoke" && (s.action || "").includes("continueOnPermanentFailure")
  );

  const paymentCommitted = CAPTURED_STATUSES.has(order.payment_status) && (order.payments || []).length > 0;
  const fulfillmentMissing =
    STUCK_FULFILLMENT_STATUSES.has(order.fulfillment_status) && (order.fulfillments || []).length === 0;
  if (paymentCommitted && fulfillmentMissing && hasContinueOnFailure) {
    return "orphaned_payment_no_fulfillment";
  }

  const hasDanglingReservation = failedSteps.some(
    (s) => s.action === "reserveInventoryStep" && s.handlerType === "invoke"
  );
  if (hasDanglingReservation && (order.items || []).length === 0) {
    return "orphaned_reservation_no_order_line";
  }

  return "ok";
}

async function getAdminToken() {
  const res = await fetch(`${BACKEND_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: ADMIN_EMAIL, password: ADMIN_PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function adminGet(token, path, params = {}) {
  const url = new URL(`${BACKEND_URL}${path}`);
  for (const [key, value] of Object.entries(params)) {
    url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Medusa ${res.status} on GET ${path}`);
  return res.json();
}

async function adminDelete(token, path) {
  const res = await fetch(`${BACKEND_URL}${path}`, {
    method: "DELETE",
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`Medusa ${res.status} on DELETE ${path}`);
  return res.json();
}

async function listRecentOrders(token, sinceIso) {
  const orders = [];
  let offset = 0;
  const limit = 100;
  while (true) {
    const data = await adminGet(token, "/admin/orders", {
      fields: "id,status,fulfillment_status,payment_status,*payments,*fulfillments,*items",
      "created_at[$gte]": sinceIso,
      limit,
      offset,
    });
    orders.push(...data.orders);
    offset += limit;
    if (offset >= data.count) return orders;
  }
}

/**
 * Placeholder hook: in a real deployment, load the failed-step trail for this
 * order's workflow transaction (for example from your own audit log of
 * { result, errors } from someWorkflow(container).run({ input, throwOnError: false })).
 * Returns [] when there is nothing on file, which classifyOrphan treats as "ok".
 */
function failedStepsForOrder(order) {
  return order._failed_steps || [];
}

export async function run() {
  const token = await getAdminToken();
  const sinceIso = new Date(Date.now() - SINCE_HOURS * 3600 * 1000).toISOString();
  const orders = await listRecentOrders(token, sinceIso);

  let reported = 0;
  let cleaned = 0;
  for (const order of orders) {
    const failedSteps = failedStepsForOrder(order);
    const outcome = classifyOrphan(order, failedSteps);
    if (outcome === "ok") continue;

    if (DELETABLE.has(outcome)) {
      const reservationId = order._dangling_reservation_id;
      console.warn(
        `Order ${order.id} classified as ${outcome}. reservation_id=${reservationId}. ${DRY_RUN ? "Would delete" : "Deleting"}`
      );
      if (!DRY_RUN && reservationId) await adminDelete(token, `/admin/reservations/${reservationId}`);
      cleaned++;
    } else {
      const record = {
        order_id: order.id,
        action: failedSteps.find((s) => s.handlerType === "invoke")?.action ?? null,
        error_message: failedSteps[0]?.message ?? null,
        classification: outcome,
        reported_at: new Date().toISOString(),
      };
      console.log(`Orphan detected: ${JSON.stringify(record)}`);
      reported++;
    }
  }

  console.log(`Done. ${reported} order(s) reported for human triage, ${cleaned} reservation(s) ${DRY_RUN ? "to delete" : "deleted"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
